oneVPL for intel gpus

evanrichter opened the pull request:
## Description of changes

fresh PR continuing from #242359, sorry for the mass-ping over there :(

ping for the people who seem interested: @Atemu @PJungkamp @J-Swift

outstanding question for @midchildan: should intel-media-sdk be marked deprecated? The [upstream repo](https://github.com/Intel-Media-SDK/MediaSDK#media-sdk-support-matrix) is archived and suggests transitioning to onevpl

also todo: Atemu says ofborg build failed for onevpl-intel-gpu on aarch-linux, but I can't find the CI run for it. marking as draft until that's re-run and I can debug it

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

pjungkamp committed:
ffmpeg: add withVPL option for libvpl

Co-authored-by: Evan Richter <evanjrichter@gmail.com>

evanrichter commented:
currently building/running the passthru tests for ffmpeg-full which is taking a long time on my machine (x86_64-linux), but trying to be as thorough as I can. I'll run nixpkgs-review next

midchildan commented:
I think it's problematic to completely remove Media SDK because that means dropping support for CPUs older than 3 years.

My preference would be to have oneVPL dispatch to Media SDK and make the Media SDK derivation private to oneVPL. Intel mentions an unspecified security issue with Media SDK, so we could make an option that allows security conscious users to disable Media SDK dispatch.

midchildan commented:
ffmpeg is the only package in Nixpkgs that uses Media SDK, so I'm assuming that removing Media SDK from the top level package set would largely go unnoticed as long as ffmpeg could dispatch to it.

PJungkamp commented:
Here's a log for the build failure of `intel-media-sdk`: [log.txt](https://github.com/NixOS/nixpkgs/files/13220691/log.txt)

As media SDK is only needed to support x86 CPUs we should only have oneVPL depend on it on x86_64. oneVPL itself should also be packaged for aarch64 as it may be used with intel dedicated graphics.

evanrichter commented:
the cause of onevpl-intel-gpu being broken on aarch64-linux is this part of upstream's cmake:

```cmake
add_library(mfx_require_sse4_properties INTERFACE)

if (CMAKE_C_COMPILER_ID MATCHES Intel)
  target_compile_options(mfx_require_sse4_properties
    INTERFACE
      $<$<PLATFORM_ID:Windows>: /arch:sse4.2>
      $<$<PLATFORM_ID:Linux>:   -xsse4.2>
    )
else()
  target_compile_options(mfx_require_sse4_properties
    INTERFACE
      # on Windows MSVC SSE4.2 supported by default
      $<$<PLATFORM_ID:Linux>:   -msse4.2>
    )
endif()
```

which goes beyond my cmake ability to patch. I agree that aarch64-linux hosts should be able to use this intel gpu driver, but obviously intel has been narrow-minded in their implementation as of now :)

I marked this driver as broken for aarch64 for now

Atemu commented on pkgs/development/libraries/oneVPL-intel-gpu/default.nix:
```diff
@@ -0,0 +1,36 @@
+    changelog = "";
+    license = [ lib.licenses.mit ];
+    platforms = lib.platforms.linux;
+    broken = stdenv.hostPlatform.isAarch64;
```
```suggestion
    broken = !stdenv.hostPlatform.isx86;
```

Atemu replied:
I'd also like to see a note why that is (cmake script adding sse4 flag); ideally in the form of a link to an upstream issue ;)

Atemu commented on pkgs/development/libraries/ffmpeg/generic.nix:
```diff
@@ -482,6 +485,9 @@ stdenv.mkDerivation (finalAttrs: {
     (enableFeature withVorbis "libvorbis")
     (enableFeature withVmaf "libvmaf")
     (enableFeature withVpx "libvpx")
+    # (enableFeature withVpl "libvpl")
+    # temporary use of optionals to avoid big rebuild
+    ] ++ optionals withVpl [ (enableFeature withVpl "libvpl") ] ++ [
```
Nit:

```suggestion
  ] ++ optionals withVpl [
    # temporary use of optionals to avoid big rebuild, removed in staging
    (enableFeature withVpl "libvpl")
  ] ++ [
```

Atemu replied:
And then you'd simply remove 488 and 491 on staging.

Atemu reviewed (approved):
ffmpeg part LGTM

evanrichter commented:
> My preference would be to have oneVPL dispatch to Media SDK and make the Media SDK derivation private to oneVPL. Intel mentions an unspecified security issue with Media SDK in their deprecation notice, so we could make an option that allows security conscious users to disable Media SDK dispatch.

I think that's a great next step but this PR is "big" enough to me (not in lines changed, just in stepping out of my comfort zone with packages that affect lots of users). I can try a draft of that change after if that's alright

Atemu commented:
Note that, while you are expected to do reasonable QA on your own, your PR does not have to be perfect. If we find out later that this breaks something unforseen, we can always revert it again (or otherwise fix it). That's one of the purposes of the unstable channel.

Obviously we'd prefer not to have to do that but it wouldn't be the end of the world either.

evanrichter commented:
I don't know why but nixpkgs-review builds a ton of seemingly unrelated packages (like systemd) and never succeeds. Should I rebase to a newer unstable checkout and try again?

I'm on x86-Linux and it seems like ofborg passed for that system

Atemu commented:
It's probably building against staging which is expected to always need a ton of rebuilds (that's what it's for).

Use the commit as basis for review rather than the merge, nixpkgs-review has a flag for that which I can never remember the name of.

evanrichter commented:
If you remember the flag for it that would be helpful. I don't see anything similar in the --help text of nixpkgs-review. Eventually this command finally passed it seems like: `nix-shell -p nixpkgs-review --run "nixpkgs-review rev -b staging HEAD"`.

Atemu commented:
If you're using `rev` mode, you need to compare your branch's head to its merge base with master (assuming it's based on master); in other words the commit your branch is "based on".

You seem to have based your branch on staging however, so you're going to need to build quite a lot of packages either way.

I usually base my branches on master/unstable for this reason. With staging you obviously need to be careful and check whether that master has made its way into the target branch yet (because [sometimes it might not have](https://github.com/NixOS/nixpkgs/pull/204198#issuecomment-1398948785)) but it makes working on the branch and reviewing _much_ easier.  
If there are merge conflicts, you'd obviously have to rebase onto staging but that rarely happens IME.

evanrichter commented:
@Atemu From your message on the last PR which I messed up by pinging everyone:

> I'd rather just target staging here

That's why I rebased to staging and changed the PR to target staging. So I should rebase my commits (back) onto master but still target staging branch? I'll read the contributing guide again with a closer eye so I don't screw up this PR also ;)

Atemu commented:
> That's why I rebased to staging and changed the PR to target staging.

Rebasing onto staging is not necessary. As long as master's/unstable's commits are in staging, you can simply change the target. GitHub only shows commits/changes in a PR that do not yet exist in the target branch.

You really ought to check whether your base has made its way to staging yet though. It happens via automation but there might be hours of delay and there can always be merge conflicts blocking those automatic merges which was the case with my ffmpeg PR snafu.  
I usually base on nixos-unstable as it's exceptionally rare for days-old master aka. nixos-unstable to not be in staging but you should always double check when you target staging.

> I should rebase my commits (back) onto master but still target staging branch?

Yes but since your current base (staging) is further ahead than master, you will have to rebase a subset (your 4 commits) or else git will try to rebase the entirety of the staging cycle ontop of master and your 4 commits ontop of that whole mess.

> I'll read the contributing guide again with a closer eye so I don't screw up this PR also ;)

I don't think the contributing guide mentions this approach. IIRC it recommends the failsafe method where you don't have to worry about master being in staging and such but isn't optimal when it comes to minimising rebuilds for working/reviewing.

There isn't much magical to this approach if you know a little about how git's model works but I've noticed that git remains a mystery to most people ;)

> I'm struggling with nixpkgs-review.

Yeah, it's unintuitive.

> If I rebase to master then force push, I can use the `pr` mode, but then I risk blowing up the thread

Even if you rebase on top of master, `pr` mode would still merge the branch into its target and use that as the "your changes" point for comparison. To avoid that, you need the alternative mode I allured to. It compares your branch's HEAD with its merge base with the target branch instead, without merging your changes with the current revision of the target.

> so I think I want to stay with `rev` mode to test locally?

Sure.

> But it's not clear how to do what you are saying. Am I trying to do a 3-way compare of HEAD, HEAD~3 (because I have 4 commits), and master?

You rebase the subset (your 4 commits) onto master. I don't know how to do that on the CLI actually because I always use magit which offers a visual commit selection for this function.

I believe this is what you'd need to to: https://stackoverflow.com/a/34927492/6513193

RaHoni commented:
Are there any updates on this? I would love to switch my desktop to NixOS but need to encode av1 with my GPU. Or can I contribute something?

pearsche commented:
yeah, it seems this died after a month..

evanrichter commented:
Feel free to use this as it is, it is working for me :) 

You can take this branch as an extra flake input and apply an overlay for just the packages I added or changed.

If they need to be updated just let me know

Atemu commented on pkgs/development/libraries/intel-media-sdk/default.nix:
```diff
@@ -18,6 +18,11 @@ stdenv.mkDerivation rec {
   ];
   nativeCheckInputs = [ gtest ];
 
+  # For forward compatibility with oneVPL.
+  # To ensure the VPL backend is dispatched at runtime, export INTEL_MEDIA_RUNTIME=ONEVPL
+  # See https://github.com/Intel-Media-SDK/MediaSDK#media-sdk-support-matrix
+  runtimeInputs = [ oneVPL ];
+
```
This means it still works on older GPUs but it uses oneVPL on newer ones, right?

evanrichter replied:
yes, it only uses oneVPL at runtime if the environment variable `INTEL_MEDIA_RUNTIME` is `ONEVPL`

Atemu reviewed (commented):
Needs a rebase for ffmpeg 6.1.

evanrichter commented on pkgs/development/libraries/ffmpeg/generic.nix:
```diff
@@ -482,6 +485,10 @@ stdenv.mkDerivation (finalAttrs: {
     (enableFeature withVorbis "libvorbis")
     (enableFeature withVmaf "libvmaf")
     (enableFeature withVpx "libvpx")
+  ] ++ optionals withVpl [
+    # temporary use of optionals to avoid big rebuild
+    (enableFeature withVpl "libvpl")
+  ] ++ [
```
it was introduced in [this commit](https://github.com/FFmpeg/FFmpeg/commit/7158f1e64d9b76afea78537a35c465447df0cff8) and these tags contain it:

- [n6.2-dev](https://github.com/FFmpeg/FFmpeg/releases/tag/n6.2-dev)
- [n6.1](https://github.com/FFmpeg/FFmpeg/releases/tag/n6.1)
- [n6.1-dev](https://github.com/FFmpeg/FFmpeg/releases/tag/n6.1-dev)
- [n6.0.1](https://github.com/FFmpeg/FFmpeg/releases/tag/n6.0.1)
- [n6.0](https://github.com/FFmpeg/FFmpeg/releases/tag/n6.0)

So I assume that means version 6 was the first

Atemu replied:
In that case, we must keep this an optionals and guard it behind versionAtLeast.

```suggestion
  ] ++ optionals (versionAtLeast version "6.0") [
    (enableFeature withVpl "libvpl")
  ] ++ [
```

evanrichter commented:
rebased to staging from earlier today, updated oneVPL to 2023.4.0, and made it only for ffmpeg >= 6.0

Atemu reviewed (approved):
Diff LGTM, can't test this without Intel hardware.

arthsmn commented:
Is this feature enabled only when explicitly overriding withVpl?

Atemu commented:
For now, yes.

arthsmn reviewed (approved):
LGTM.

Atemu commented:
Could you check by how much this bloats the ffmpeg closure? If it's not too much, I think we can merge.

RaHoni commented:
> Could you check by how much this bloats the ffmpeg closure? If it's not too much, I think we can merge.

How could I do that. I would be willing to do that but haven't done it before.

Atemu commented:
nix path-info -Sh

RaHoni commented:
Without overriding withVPL
```
❯ nix path-info -Sh  .#ffmpeg
/nix/store/2n5aqxbl1a9v1azlm48ffdrl19a8crsb-ffmpeg-6.1-bin       490.6M
/nix/store/rbfmrvqjg3jansapsw2f2hz0lyil8qyv-ffmpeg-6.1-man         1.3M
```

CobaltCause commented:
Upstream just renamed the repository and changed the version scheme: https://github.com/intel/libvpl/releases/tag/v2.10.1

evanrichter commented:
thanks for the heads up!

I can see about the pkgs/by-name scheme as well..

CobaltCause commented:
Since OBS Studio needs `libvpl` (formerly oneVPL) to build now and this PR targets `staging`, would you consider either:

1. Opening a new PR targeting `master` just to package `libvpl` and not touching other packages
2. Being okay with the `libvpl` package definition from https://github.com/NixOS/nixpkgs/pull/275544 getting merged

And then rebasing this PR on top of one of those changes?

evanrichter commented:
@CobaltCause I'd rather rebase after your PR makes it to staging. I don't feel like going through the headache of switching target branches or making more PRs at this time.

I did just check out your libvpl package definition. It's surprising that you don't need the laundry list of X and Wayland buildInputs that I needed to wrestle in! I'd probably go with your definition at this time since it seems like upstream is changing things quickly and I looked at this over a month ago now.

CobaltCause commented:
> @CobaltCause I'd rather rebase after your PR makes it to staging. I don't feel like going through the headache of switching target branches or making more PRs at this time.

Sounds good to me!

> I did just check out your libvpl package definition. It's surprising that you don't need the laundry list of X and Wayland buildInputs that I needed to wrestle in!

Yeah, the package definition in my PR disables the tools, i.e. the programs that otherwise get installed to `$out/bin`. Which I guess raises the question, do you need/want those to exist? It seems like the main function of the package is just as a library anyway.

Atemu commented:
> want those to exist

Note that we have the possibility to do multiple outputs; we can have both the tools and the library without polluting users' systems.

evanrichter commented:
yeah I think most will just use the library, including me :)

@Atemu is that a simple addition? `outputs = ["bin" "dev" "blah"];` has always confused me and seems to not do what I expect

Atemu commented:
It's been a while since I've done it but it should be a matter of declaring multiple outputs (default output to be installed first; `bin` in most cases) and then passing them as configureFlags via `placeholder`.

evanrichter committed:
onevpl-intel-gpu: init at 23.4.3

evanrichter commented:
I rebased on top of the merge of the obs-studio update (by @CobaltCause), and am now using their libvpl package.

I made the new package `oneVPL-intel-gpu` follow the new pkgs/by-name convention, and it builds locally for me with `nix-build . -A oneVPL-intel-gpu`, so the [CI parseable check failure](https://github.com/NixOS/nixpkgs/actions/runs/7632197814/job/20791863781?pr=264621#step:5:13) is confusing to me. Is that check wrong, or did I do by-name incorrectly?

evanrichter commented:
I rebased on a very current staging just now to resolve a simple conflict. I haven't built these packages yet because I don't feel like building the entire world at the moment. I'll wait a couple days for the cache and try again

RaHoni commented:
As I haven't check this before an update 😒 I am currently rebuilding it.

evanrichter commented:
> As I haven't check this before an update 😒 I am currently rebuilding it.

My other branch with the same name but "-cb" at the end has everything except the packages I changed already in the cache.

RaHoni commented:
@evanrichter Thanks for the Tip since in the normal branch blueZ won't compile I switched to the Cherry Picked version. While I am still not abel to use av1 qsv with ffmpeg the compile of your Branch still works.

evanrichter commented:
> @evanrichter Thanks for the Tip since in the normal branch blueZ won't compile I switched to the Cherry Picked version. While I am still not abel to use av1 qsv with ffmpeg the compile of your Branch still works.

did you override the compile options? Like this:

```nix
ffmpeg-full.override {
  withMfx = false;
  withVpl = true;
};
```

and maybe you need some environment variables too, I forget:
```nix
  environment.sessionVariables = {
    INTEL_MEDIA_RUNTIME= "ONEVPL";
    LIBVA_DRIVER_NAME = "iHD";
    ONEVPL_SEARCH_PATH = lib.strings.makeLibraryPath (with pkgs; [oneVPL-intel-gpu]);
  };
```

Atemu commented:
@evanrichter that should probably be documented somewhere...

RaHoni commented:
> and maybe you need some environment variables too, I forget:
> 
> ```nix
>   environment.sessionVariables = {
>     INTEL_MEDIA_RUNTIME= "ONEVPL";
>     LIBVA_DRIVER_NAME = "iHD";
>     ONEVPL_SEARCH_PATH = lib.strings.makeLibraryPath (with pkgs; [oneVPL-intel-gpu]);
>   };
> ```

Yeah I was overriding the build flags but I haven't set the envoirment variables. Thanks.

evanrichter commented:
> @evanrichter that should probably be documented somewhere...

where should it go you think? nixos.wiki?

RaHoni commented:
I would expect it in https://nixos.wiki/wiki/Intel_Graphics or https://nixos.wiki/wiki/Accelerated_Video_Playback if you also need it for playback.

fricklerhandwerk commented:
@RaHoni writing a module for NixOS is always the better way of doing things with NixOS. But this PR just adds a package, which is fine on its own and should simply be merged if reviewers approve.

IIRC, the last state of discussion at the docs team on NixOS modules was: infrastructure for cross-referencing is nonexistent, but still put documentation into the module itself if possible. In this case I can imagine listing which packages are relevant in the `hardware.opengl` module.

We're still lacking maintainers for NixOS documentation, and someone needs to invest some brainpower and elbow grease into establishing practically viable structure and building automation. The (soon new) Wiki is still supposed to be dumping ground, not the single source of truth.

RyanGibb commented:
> Diff LGTM, can't test this without Intel hardware.

@Atemu I can confirm that this is working on the N100 Alder Lake.

Atemu commented:
@ofBorg eval

RyanGibb commented:
I have a final comment on:

> ```nix
>   environment.sessionVariables = {
>     INTEL_MEDIA_RUNTIME= "ONEVPL";
>     LIBVA_DRIVER_NAME = "iHD";
>     ONEVPL_SEARCH_PATH = lib.strings.makeLibraryPath (with pkgs; [oneVPL-intel-gpu]);
>   };
> ```

I didn't require `INTEL_MEDIA_RUNTIME` or `LIBVA_DRIVER_NAME`: they were picked up automatically.

To understand `ONEVPL_SEARCH_PATH`, it looks like `libvpl` gets a runtime implementation (according too)[https://spec.oneapi.io/onevpl/latest/programming_guide/VPL_prg_session.html#onevpl-dispatcher]:
1. Directories provided by the environment variable LD_LIBRARY_PATH.
2. Default path /lib, then /usr/lib or /lib64, and then /usr/lib64 on some 64 bit OSs. On Debian: /usr/lib/x86_64-linux-gnu.
3. Current working directory.
4. For backward compatibility with older spec versions, dispatcher also checks user-defined search folders which are provided by ONEVPL_SEARCH_PATH environmental variable.

So instead of using this environment variable, perhaps a wrapper around `ffmpeg` to set `LD_LIBRARY_PATH` would be appropriate?

Atemu commented:
If anything, libvpl should be patched to also look in `/run/opengl-driver/lib/`.

We will not set LD_LIBRARY_PATH on ffmpeg.

RyanGibb commented:
> If anything, libvpl should be patched to also look in `/run/opengl-driver/lib/`.

Why `/run/opengl-driver/lib/`?

RyanGibb commented:
I've made a PR to upstream the libvpl patch here: https://github.com/NixOS/nixpkgs/pull/291559

SuperSandro2000 commented on pkgs/by-name/on/oneVPL-intel-gpu/package.nix:
```diff
@@ -0,0 +1,38 @@
+  meta = {
+    description = "oneAPI Video Processing Library Intel GPU implementation";
+    homepage = "https://github.com/oneapi-src/oneVPL-intel-gpu";
+    changelog = "";
```
```suggestion
    changelog = "https://github.com/oneapi-src/oneVPL-intel-gpu/releases/tag/${srv.rev}";
```

evanrichter replied:
I'll apply this, thanks!

evanrichter replied:
done in latest push

SuperSandro2000 commented on pkgs/development/libraries/intel-media-sdk/default.nix:
```diff
@@ -27,6 +27,11 @@ stdenv.mkDerivation rec {
+  # For forward compatibility with oneVPL.
+  # To ensure the VPL backend is dispatched at runtime, export INTEL_MEDIA_RUNTIME=ONEVPL
+  # See https://github.com/Intel-Media-SDK/MediaSDK#media-sdk-support-matrix
+  runtimeInputs = [ libvpl ];
```
Can we use patchelf instead of LD_LIBRARY_PATH?

evanrichter replied:
oh good point, I can try that

evanrichter replied:
I keep coming back to this, hoping more of staging is present in cache.nixos.org, otherwise dependencies are failing to build for me... today it's progressing farther.

my change looks like this, removing runtimeInputs:
```nix
  postFixup = ''
    patchelf --add-needed ${libvpl}/lib/libvpl.so
  '';
```

I don't remember where I found runtimeInputs, and it doesn't seem documented. What is the actual difference between runtimeInputs and patchelf?

Atemu replied:
Staging is intentionally not cached or even built for the most part.

My recommendation always is to base your staging PRs on the merge-base between master and staging. That way you can test things with the cache of master for everything but the thing you're touching and it should still apply cleanly to staging.

I don't remember the exact flag name but I think this would technically be better as it'd allow grafting:

```Nix
  postFixup = ''
    patchelf --add-needed libvpl.so --add-rpath ${libvpl}/lib/libvpl.so
  '';
```

evanrichter replied:
reading the intel-media-sdk compatibility chart again I'm pretty sure I needed to patchelf onevpl-intel-gpu's libmfx-gen.so onto intel-media-sdk's libmfx.so so it can find it at runtime if directed. [This line in intel-media-sdk](https://github.com/Intel-Media-SDK/MediaSDK/blob/7a72de33a15d6e7cdb842b12b901a003f7154f0a/api/mfx_dispatch/linux/mfxloader.cpp#L194) is clearly trying to load libmfx-gen.so which is not provided by libvpl, but by onevpl-intel-gpu.

@Atemu This raises the closure size of intel-media-sdk (and ffmpeg, if withMfx or withVpl are set) by 10.6MB.

I'm not exactly sure how to test this fallback mechanism in isolation, but I've tried the following with success:
1. use this PR's intel-media-sdk and ffmpeg-full (with no overrides)
2. use ffmpeg to encode using a qsv backend without `INTEL_MEDIA_RUNTIME` set
3. use ffmpeg to encode using a qsv backend with `INTEL_MEDIA_RUNTIME=ONEVPL`
4. check with `intel_gpu_top` that it is in fact using the gpu for both encodes

evanrichter replied:
using patchelf in the latest revision, btw

after looking more, I don't think `runtimeInputs` is even read by `mkDerivation`

evanrichter replied:
I've removed the patch for intel-media-sdk based on my reasoning above. I think if there's a clear need for it, it can easily be added back in another PR to make incremental improvement

Atemu replied:
> EDIT 2: using libvpl from #291559 and `hardware.opengl.extraPackages = [ pkgs.onevpl-intel-gpu ];` I can use this PR's ffmpeg-full `withVpl = true;` with an **unmodified intel-media-sdk**.

This is how it should work. `onevpl-intel-gpu` is a driver package and should be installed to the global driver path by the user.

That's how all other hardware-dependant accelerator APIs work in Nixpkgs.

SuperSandro2000 commented on pkgs/by-name/on/oneVPL-intel-gpu/package.nix:
```diff
@@ -0,0 +1,38 @@
+    changelog = "";
+    license = [ lib.licenses.mit ];
+    platforms = lib.platforms.linux;
+    # CMake adds x86 specific compiler flags in <source>/builder/FindGlobals.cmake
+    # NOTE: https://github.com/oneapi-src/oneVPL-intel-gpu/issues/303
+    broken = !stdenv.hostPlatform.isx86;
```
Did you try to just remove the msse4.2 flag?

evanrichter replied:
I did... but the dialect of cmake in that repo is beyond my capability and free time for now

Atemu commented on pkgs/development/libraries/ffmpeg/generic.nix:
```diff
@@ -92,6 +92,7 @@
 , withVmaf ? withFullDeps && !stdenv.isAarch64 && lib.versionAtLeast version "5" # Netflix's VMAF (Video Multi-Method Assessment Fusion)
 , withVoAmrwbenc ? withFullDeps && withVersion3 # AMR-WB encoder
 , withVorbis ? withHeadlessDeps # Vorbis de/encoding, native encoder exists
+, withVpl ? false # Hardware acceleration via intel libvpl
```
How much extra closure size would this be if you enabled it now that the driver bit is impure?

evanrichter replied:
showing just the different lines:

```
# withMfx
/nix/store/556zw392nqnc0r8fg9n7yid1w920zd80-intel-media-sdk-23.2.2                 	  28.8M	  85.8M
/nix/store/mjndrx4agd3l69z4l0b98vv18dfsn9kb-ffmpeg-full-6.1-data                   	 264.9K	 264.9K
/nix/store/ndwfmpylljb6hjcpy6iz4557akc9ffmp-ffmpeg-full-6.1-lib                    	  27.7M	   1.3G
/nix/store/9ryljr6yh38sssfg4ajy3pv1c089xwzf-ffmpeg-full-6.1-bin                    	 727.2K	   1.3G

# withVpl
/nix/store/d8w5qfswmgxcjqwnmqw2v9r8amrvdlpb-xgcc-13.2.0-libgcc                     	 155.8K	 155.8K
/nix/store/vqvbn2z8wyrjwvayjb2vy5krhh1kis9b-libunistring-1.1                       	   1.7M	   1.7M
/nix/store/krqp9wj3rgalmqv04y0sqw987mxsnddn-libidn2-2.3.7                          	 352.7K	   2.1M
/nix/store/ksk3rnb0ljx8gngzk19jlmbjyvac4hw6-glibc-2.38-44                          	  28.9M	  31.1M
/nix/store/4vzal97iq3dmrgycj8r0gflrh51p8w1s-bash-5.2p26                            	   1.5M	  32.6M
/nix/store/alxp4kb0dgfczqwd7xfrpflhq50w0wck-gcc-13.2.0-libgcc                      	 155.8K	 155.8K
/nix/store/pp0jsd045xvfsz60kpbkfxbs9pbpk8z5-gcc-13.2.0-lib                         	   8.7M	  39.9M
/nix/store/b7gaf7alc4qjymv1mz79cshsayf1w20s-libvpl-2.10.1                          	   1.1M	  42.5M
/nix/store/vcyijj6xrw3acjphyb0sji1klz4xyslf-ffmpeg-full-6.1-data                   	 264.9K	 264.9K
/nix/store/smhwvfrq7j5khsk8761qiyryn5lvp7qw-ffmpeg-full-6.1-lib                    	  27.8M	   1.3G
/nix/store/rhmmr863zwlrxf8pvgfixml63zl87iwd-ffmpeg-full-6.1-bin                    	 727.2K	   1.3G
```

and when I change to `withVpl` for my full system closure it's a 26 M improvement:
```
<<< /run/current-system
>>> /nix/store/r0frim4jr544rkh0fxyk7rsbd1hmqr5z-nixos-system-coffee-23.11.20240308.2be119a
Version changes:
[C.]  #1  bash  5.2-p15 x2, 5.2-p21 x2, 5.2p26 -> 5.2-p15 x2, 5.2-p21 x2, 5.2p26 x2
Added packages:
[A.]  #1  libvpl  2.10.1
Removed packages:
[R.]  #1  intel-media-sdk  23.2.2
Closure size: 2763 -> 2764 (24 paths added, 23 paths removed, delta +1, disk usage -26.2MiB).
```

Atemu replied:
Neat! Once GPUs only supported by MFX are sufficiently legacy (or if ffmpeg were to support both), I see nothing blocking its enablement then.

evanrichter replied:
thanks for all your help getting this PR merged! 😄

huantianad commented:
In order to have the most optimal setup for hardware acceleration with a recent Intel CPU then, is `intel-media-driver` the only package I should be adding to my open gl package list?

evanrichter commented:
> In order to have the most optimal setup for hardware acceleration with a recent Intel CPU then, is `intel-media-driver` the only package I should be adding to my open gl package list?

You actually just want `onevpl-intel-gpu` instead:

```nix
  hardware.opengl = {
    enable = true;
    extraPackages = with pkgs; [
      onevpl-intel-gpu
    ];
  };
```

and you want to override your ffmpeg to use vpl:
```nix
        ffmpeg-full = prev.ffmpeg-full.override {
          withVpl = true; withMfx = false;
        };
```

and that should be it, just check that this pr is included in your nixpkgs too: #291559

huantianad commented:
Ah OK, should this work for any programs that are dependent on intel media driver or media sdk as well? I did see that there was mention of it having backwards compat so it would.

programotuojes commented:
This probably depends on the use case, but for transcoding in Jellyfin, `intel-media-driver` was still required for me. Without it, ffmpeg errored out:

```
[AVHWDeviceContext @ 0x1ca9ac0] Failed to initialise VAAPI connection: -1 (unknown libva error).
Device creation failed: -5.
Failed to set value 'vaapi=va:,kernel_driver=i915,driver=iHD' for option 'init_hw_device': Input/output error
Error parsing global options: Input/output error
```

Additionally, since I wanted to enable regular tone mapping in Jellyfin (the VPP tone mapping was too dark for me), I had to add both `intel-compute-runtime` and `intel-ocl`. Only adding `intel-ocl` resulted in this error:

```
[AVHWDeviceContext @ 0x1678d00] Failed to get number of OpenCL platforms: -1001.
Device creation failed: -19.
Failed to set value 'opencl=ocl@va' for option 'init_hw_device': No such device
Error parsing global options: No such device
```

This is the config I use while this PR is still [underway](https://nixpk.gs/pr-tracker.html?pr=264621). Mostly copied from [here](https://github.com/RyanGibb/nixos/blob/0dbb10a2e7deee8b76129c9f32c07ad7bfb555a5/hosts/elephant/default.nix#L73).

```nix
boot.kernelParams = [ "i915.enable_guc=3" ];

nixpkgs.config.packageOverrides = prev: {
  ffmpeg_6 = prev.ffmpeg_6.overrideAttrs (old: rec {
    configureFlags =
      # Remove deprecated Intel Media SDK support
      (builtins.filter (e: e != "--enable-libmfx") old.configureFlags)
      # Add Intel VPL support
      ++ [ "--enable-libvpl" ];
    buildInputs = old.buildInputs ++ [
      # VPL dispatcher
      pkgs-unstable.libvpl
    ];
  });
};

hardware.opengl = {
  enable = true;
  extraPackages = with pkgs; [
    (pkgs.callPackage ../../pkgs/onevpl-intel-gpu.nix { })
    intel-media-driver
    intel-compute-runtime
    intel-ocl
  ];
};
```

Atemu commented:
@programotuojes jellyfin-ffmpeg is separate from ffmpeg. It's a variant of ffmpeg_6-full. If you need VPL in jellyfin, you need to override `jellyfin-ffmpeg`. Please also note that you should also use overlays rather than packageOverrides.

huantianad commented:
> Ah OK, should this work for any programs that are dependent on intel media driver or media sdk as well? I did see that there was mention of it having backwards compat so it would.

In addition, I assume that it will use the same entry point as the older intel media driver and sdk, so `LIBVA_DRIVER_NAME = "iHD"` should still be used?

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/jellyfin-docker-container-on-intel-gpu-opencl-devices-not-found/42119/1

TimoVerbrugghe commented:
> This probably depends on the use case, but for transcoding in Jellyfin, `intel-media-driver` was still required for me. Without it, ffmpeg errored out:
> 
> ```
> [AVHWDeviceContext @ 0x1ca9ac0] Failed to initialise VAAPI connection: -1 (unknown libva error).
> Device creation failed: -5.
> Failed to set value 'vaapi=va:,kernel_driver=i915,driver=iHD' for option 'init_hw_device': Input/output error
> Error parsing global options: Input/output error
> ```
> 
> Additionally, since I wanted to enable regular tone mapping in Jellyfin (the VPP tone mapping was too dark for me), I had to add both `intel-compute-runtime` and `intel-ocl`. Only adding `intel-ocl` resulted in this error:
> 
> ```
> [AVHWDeviceContext @ 0x1678d00] Failed to get number of OpenCL platforms: -1001.
> Device creation failed: -19.
> Failed to set value 'opencl=ocl@va' for option 'init_hw_device': No such device
> Error parsing global options: No such device
> ```
> 
> This is the config I use while this PR is still [underway](https://nixpk.gs/pr-tracker.html?pr=264621). Mostly copied from [here](https://github.com/RyanGibb/nixos/blob/0dbb10a2e7deee8b76129c9f32c07ad7bfb555a5/hosts/elephant/default.nix#L73).
> 
> ```nix
> boot.kernelParams = [ "i915.enable_guc=3" ];
> 
> nixpkgs.config.packageOverrides = prev: {
>   ffmpeg_6 = prev.ffmpeg_6.overrideAttrs (old: rec {
>     configureFlags =
>       # Remove deprecated Intel Media SDK support
>       (builtins.filter (e: e != "--enable-libmfx") old.configureFlags)
>       # Add Intel VPL support
>       ++ [ "--enable-libvpl" ];
>     buildInputs = old.buildInputs ++ [
>       # VPL dispatcher
>       pkgs-unstable.libvpl
>     ];
>   });
> };
> 
> hardware.opengl = {
>   enable = true;
>   extraPackages = with pkgs; [
>     (pkgs.callPackage ../../pkgs/onevpl-intel-gpu.nix { })
>     intel-media-driver
>     intel-compute-runtime
>     intel-ocl
>   ];
> };
> ```

@programotuojes: it's interesting that you say that you had to both add intel-ocl and intel-compute-runtime as opengl extra packages. I tried this as well because I'm trying to run jellyfin opencl tone mapping as well (however inside of a docker container, not as a nixos service defined in configuration.nix). Added your PR, intel-ocl and intel-compute-runtime but jellyfin tone mapping still failing for me with the same error.

Additionally, if I do clinfo on the nixos machine, it seems to detect openCL for the CPU, but not on the GPU (instead of f.e. on my N5105 machine where it detects 2 opencl platforms, 1 CPU and GPU).

When you do clinfo, your machine with the PR applied shows 2 opencl platforms?

N100 clinfo output:

```
[nix-shell:/home/odd]# clinfo
Number of platforms                               1
  Platform Name                                   Intel(R) OpenCL
  Platform Vendor                                 Intel(R) Corporation
  Platform Version                                OpenCL 1.2 LINUX
  Platform Profile                                FULL_PROFILE
  Platform Extensions                             cl_khr_icd cl_khr_global_int32_base_atomics cl_khr_global_int32_extended_atomics cl_khr_local_int32_base_atomics cl_khr_local_int32_extended_atomics cl_khr_byte_addressable_store cl_khr_depth_images cl_khr_3d_image_writes cl_intel_exec_by_local_thread cl_khr_spir cl_khr_fp64
  Platform Extensions function suffix             INTEL

  Platform Name                                   Intel(R) OpenCL
Number of devices                                 1
  Device Name                                     Intel(R) N100
  Device Vendor                                   Intel(R) Corporation
  Device Vendor ID                                0x8086
  Device Version                                  OpenCL 1.2 (Build 475)
  Driver Version                                  1.2.0.475
  Device OpenCL C Version                         OpenCL C 1.2
  Device Type                                     CPU
  Device Profile                                  FULL_PROFILE
  Device Available                                Yes
  Compiler Available                              Yes
  Linker Available                                Yes
  Max compute units                               4
  Max clock frequency                             0MHz
  Device Partition                                (core)
    Max number of sub-devices                     4
    Supported partition types                     by counts, equally, by names (Intel)
    Supported affinity domains                    (n/a)
  Max work item dimensions                        3
  Max work item sizes                             8192x8192x8192
  Max work group size                             8192
  Preferred work group size multiple (kernel)     128
  Preferred / native vector sizes
    char                                                 1 / 32
    short                                                1 / 16
    int                                                  1 / 8
    long                                                 1 / 4
    half                                                 0 / 0        (n/a)
    float                                                1 / 8
    double                                               1 / 4        (cl_khr_fp64)
  Half-precision Floating-point support           (n/a)
  Single-precision Floating-point support         (core)
    Denormals                                     Yes
    Infinity and NANs                             Yes
    Round to nearest                              Yes
    Round to zero                                 No
    Round to infinity                             No
    IEEE754-2008 fused multiply-add               No
    Support is emulated in software               No
    Correctly-rounded divide and sqrt operations  No
  Double-precision Floating-point support         (cl_khr_fp64)
    Denormals                                     Yes
    Infinity and NANs                             Yes
    Round to nearest                              Yes
    Round to zero                                 Yes
    Round to infinity                             Yes
    IEEE754-2008 fused multiply-add               Yes
    Support is emulated in software               No
  Address bits                                    64, Little-Endian
  Global memory size                              12512915456 (11.65GiB)
  Error Correction support                        No
  Max memory allocation                           3128228864 (2.913GiB)
  Unified memory for Host and Device              Yes
  Minimum alignment for any data type             128 bytes
  Alignment of base address                       1024 bits (128 bytes)
  Global Memory cache type                        Read/Write
  Global Memory cache size                        524288 (512KiB)
  Global Memory cache line size                   64 bytes
  Image support                                   Yes
    Max number of samplers per kernel             480
    Max size for 1D images from buffer            195514304 pixels
    Max 1D or 2D image array size                 2048 images
    Max 2D image size                             16384x16384 pixels
    Max 3D image size                             2048x2048x2048 pixels
    Max number of read image args                 480
    Max number of write image args                480
  Local memory type                               Global
  Local memory size                               32768 (32KiB)
  Max number of constant args                     480
  Max constant buffer size                        131072 (128KiB)
  Max size of kernel argument                     3840 (3.75KiB)
  Queue properties
    Out-of-order execution                        Yes
    Profiling                                     Yes
    Local thread execution (Intel)                Yes
  Prefer user sync for interop                    No
  Profiling timer resolution                      1ns
  Execution capabilities
    Run OpenCL kernels                            Yes
    Run native kernels                            Yes
    SPIR versions                                 1.2
  printf() buffer size                            1048576 (1024KiB)
  Built-in kernels                                (n/a)
  Device Extensions                               cl_khr_icd cl_khr_global_int32_base_atomics cl_khr_global_int32_extended_atomics cl_khr_local_int32_base_atomics cl_khr_local_int32_extended_atomics cl_khr_byte_addressable_store cl_khr_depth_images cl_khr_3d_image_writes cl_intel_exec_by_local_thread cl_khr_spir cl_khr_fp64

NULL platform behavior
  clGetPlatformInfo(NULL, CL_PLATFORM_NAME, ...)  Intel(R) OpenCL
  clGetDeviceIDs(NULL, CL_DEVICE_TYPE_ALL, ...)   Success [INTEL]
  clCreateContext(NULL, ...) [default]            Success [INTEL]
  clCreateContextFromType(NULL, CL_DEVICE_TYPE_DEFAULT)  Success (1)
    Platform Name                                 Intel(R) OpenCL
    Device Name                                   Intel(R) N100
  clCreateContextFromType(NULL, CL_DEVICE_TYPE_CPU)  Success (1)
    Platform Name                                 Intel(R) OpenCL
    Device Name                                   Intel(R) N100
  clCreateContextFromType(NULL, CL_DEVICE_TYPE_GPU)  No devices found in platform
  clCreateContextFromType(NULL, CL_DEVICE_TYPE_ACCELERATOR)  No devices found in platform
  clCreateContextFromType(NULL, CL_DEVICE_TYPE_CUSTOM)  No devices found in platform
  clCreateContextFromType(NULL, CL_DEVICE_TYPE_ALL)  Success (1)
    Platform Name                                 Intel(R) OpenCL
    Device Name                                   Intel(R) N100

ICD loader properties
  ICD loader Name                                 OpenCL ICD Loader
  ICD loader Vendor                               OCL Icd free software
  ICD loader Version                              2.3.2
  ICD loader Profile                              OpenCL 3.0
```

TimoVerbrugghe commented:
Nevermind, found the problem.

Apparently there is an issue with intel-compute-runtime (which provides opencl libraries for intel GPU) on the latest 6.8 linux kernel: https://github.com/intel/compute-runtime/issues/710

I was running that kernel and that’s why the GPU did not show up under clinfo and opencl didn’t work in the jellyfin container.

Temporarily downgrading to the 6.7 kernel fixed the issue.

RyanGibb commented:
Upgrading to 24.05 has caused hardware transcoding to break for me.

```
/nix/store/ilsa2b1fgfsdhamg3ga0a5x5jljb5841-jellyfin-ffmpeg-6.0.1-6-bin/bin/ffmpeg -init_hw_device vaapi=va:/dev/dri/renderD128 -init_hw_device qsv=qs@va -hwaccel qsv -hwaccel_output_format qsv -i input.mp4 -c:v h264_qsv output.mp4#
ffmpeg version 6.0.1-Jellyfin Copyright (c) 2000-2023 the FFmpeg developers
  built with gcc 13.2.0 (GCC)
  configuration: --disable-static --prefix=/nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-jellyfin-ffmpeg-6.0.1-6 --target_os=linux --arch=x86_64 --pkg-config=pkg-config --enable-gpl --enable-version3 --disable-nonfree --disable-static --enable-shared --enable-pic --disable-thumb --disable-small --enable-runtime-cpudetect --enable-gray --enable-swscale-alpha --enable-hardcoded-tables --enable-safe-bitstream-reader --enable-pthreads --disable-w32threads --disable-os2threads --enable-network --enable-pixelutils --datadir=/nix/store/p6cyy4la6n66d134z25n1d4g0q4dc91x-jellyfin-ffmpeg-6.0.1-6-data/share/ffmpeg --enable-ffmpeg --enable-ffplay --enable-ffprobe --bindir=/nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-jellyfin-ffmpeg-6.0.1-6-bin/bin --enable-avcodec --enable-avdevice --enable-avfilter --enable-avformat --enable-avutil --enable-postproc --enable-swresample --enable-swscale --libdir=/nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-jellyfin-ffmpeg-6.0.1-6-lib/lib --incdir=/nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-jellyfin-ffmpeg-6.0.1-6-dev/include --enable-doc --enable-htmlpages --enable-manpages --mandir=/nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-jellyfin-ffmpeg-6.0.1-6-man/share/man --enable-podpages --enable-txtpages --docdir=/nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-jellyfin-ffmpeg-6.0.1-6-doc/share/doc/ffmpeg --enable-alsa --enable-libaom --disable-appkit --enable-libass --disable-audiotoolbox --disable-avfoundation --enable-avisynth --enable-libbluray --enable-libbs2b --enable-bzlib --enable-libcaca --enable-libcelt --enable-chromaprint --enable-libcodec2 --disable-coreimage --enable-cuda --enable-cuda-llvm --enable-cuvid --enable-libdav1d --enable-libdc1394 --enable-libdrm --disable-libfdk-aac --enable-ffnvcodec --enable-libflite --enable-fontconfig --enable-libfontconfig --enable-libfreetype --enable-frei0r --enable-libfribidi --enable-libgme --enable-gnutls --enable-libgsm --enable-iconv --enable-libjack --enable-libjxl --enable-ladspa --enable-lzma --enable-libmfx --enable-libmodplug --enable-libmp3lame --enable-libmysofa --enable-nvdec --enable-nvenc --enable-openal --enable-opencl --enable-libopencore-amrnb --enable-libopencore-amrwb --enable-opengl --enable-libopenh264 --enable-libopenjpeg --enable-libopenmpt --enable-libopus --enable-libplacebo --enable-libpulse --enable-librav1e --enable-librtmp --enable-libsmbclient --enable-sdl2 --enable-libshaderc --enable-libsoxr --enable-libspeex --enable-libsrt --enable-libssh --enable-librsvg --enable-libsvtav1 --disable-libtensorflow --enable-libtheora --enable-libv4l2 --enable-v4l2-m2m --enable-vaapi --enable-vdpau --disable-libvpl --disable-videotoolbox --enable-libvidstab --enable-libvmaf --enable-libvo-amrwbenc --enable-libvorbis --enable-libvpx --enable-vulkan --enable-libwebp --enable-libx264 --enable-libx265 --enable-libxavs --enable-libxcb --enable-libxcb-shape --enable-libxcb-shm --enable-libxcb-xfixes --enable-xlib --enable-libxml2 --enable-libxvid --enable-libzimg --enable-zlib --enable-libzmq --disable-debug --enable-optimizations --disable-extra-warnings --disable-stripping --extra-version=Jellyfin --disable-ptx-compression
  libavutil      58.  2.100 / 58.  2.100
  libavcodec     60.  3.100 / 60.  3.100
  libavformat    60.  3.100 / 60.  3.100
  libavdevice    60.  1.100 / 60.  1.100
  libavfilter     9.  3.100 /  9.  3.100
  libswscale      7.  1.100 /  7.  1.100
  libswresample   4. 10.100 /  4. 10.100
  libpostproc    57.  1.100 / 57.  1.100
[AVHWDeviceContext @ 0xbff280] Error initializing an MFX session: -3.
Device creation failed: -1313558101.
Failed to set value 'qsv=qs@va' for option 'init_hw_device': Unknown error occurred
Error parsing global options: Unknown error occurred
```

Has anyone else experienced this?

eclairevoyant commented:
@RyanGibb please create a new issue with config details

RyanGibb commented:
Ah, looks like [this](https://github.com/NixOS/nixpkgs/commit/ec34e3d7407c7a0f5054277941f6cb7cb41002a1#diff-ab5748dc9567516fefba8344056b51ec1866adeace380f46e58a7af3d619ea22L20995-L20997) commit changed `ffmpeg_6-full` to not override 

> This is the config I use while this PR is still [underway](https://nixpk.gs/pr-tracker.html?pr=264621). Mostly copied from [here](https://github.com/RyanGibb/nixos/blob/0dbb10a2e7deee8b76129c9f32c07ad7bfb555a5/hosts/elephant/default.nix#L73).

Changing overlay to:

```
  nixpkgs.config.packageOverrides = prev: {
    jellyfin-ffmpeg = prev.jellyfin-ffmpeg.overrideAttrs (_: {
      withVpl = true;
    });
  };
  ```
  
Fixes this. But with https://github.com/NixOS/nixpkgs/pull/315425 we can remove the overlay and use VA-API as a hardware transcoding method (e.g. in jellfin's settings), which will use oneVPL under the hood.
